### PR TITLE
feat: structured logging

### PR DESCRIPTION
**Pull request by johnxnguyen**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

Logging plain strings can be inconsistent and unstructured, making it hard to read them in large log files. Unstructured logs are also hard to parse and process.

### Solutions

Add the basic for structured logging by introducing `LogConvertible`, a protocol that simply requires a `logDescription` string. The idea is that we can create structured logs using data structures that can encode their data into a string. 

To provide an example, I've added a `Flow` to track the progress of particular code paths. The `Flow` provides a simple interface that conceals the noise of generating structured logs. For instance:

```swift
func mlsCreateGroup() throws {
  let flow = Flow(tag: "mls", name: "createGroup")

  // Records a "FLOW: {"name": "createGroup", "event":"start"}" log tagged with "mls"
  flow.start()

  ...

  // Records a "FLOW: {"name": "createGroup", "event":"checkpoint - group id is 123"}" log tagged with "mls"
  flow.checkpoint(description: "group id is 123)

  do {
    try ...

    // Records a "FLOW: {"name": "createGroup", "event":"success"}" log tagged with "mls"
    flow.succeed()
  } catch {
    // Records a "FLOW: {"name": "createGroup", "event":"fail - error description"}" log tagged with "mls"
    flow.fail(error)
  }

}

```

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: wire-ios-system/Source/Logging/Flow.swift [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+    func encode(to encoder: Encoder) throws {
+      var container = encoder.singleValueContainer()
+
+      switch self {
+      case .start:
+        try container.encode("start")
+
+      case .checkpoint(let description):
+        try container.encode("checkpoint: \(description)")
```
Comment (johnxnguyen) [inline wire-ios-system/Source/Logging/Flow.swift]:
```suggestion
        try container.encode("checkpoint - \(description)")
```

File: wire-ios-system/Source/Logging/Flow.swift [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+      case .completion(.success):
+        try container.encode("success")
+
+      case .completion(.failure(let error)):
+        try container.encode("failure: \(String(describing: error))")
```
Comment (johnxnguyen) [inline wire-ios-system/Source/Logging/Flow.swift]:
```suggestion
        try container.encode("failure - \(String(describing: error))")
```

**Comment by netbe** [comment]
@johnxnguyen is it similar to traces? Do you have a example of what it looks like on datadog for example?

**Review by agisilaos — APPROVED** [review]

**Comment by johnxnguyen** [comment]
> @johnxnguyen is it similar to traces? Do you have a example of what it looks like on datadog for example?

@netbe  Traces? No example yet but I will try to create one for you.

**Review by netbe — APPROVED** [review]
